### PR TITLE
[Merged by Bors] - feat(Logic/Function/FromTypes): Add curried heterogeneous function types

### PR DESCRIPTION
**Pull request by Shamrock-Frost**
See #10389 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Shamrock-Frost** [comment]
@eric-wieser It is possible to put `↿f` as the body of `uncurry₃`, and I'll change that if needed, but it's not possible to change the type signature of ` uncurry_3_eq_uncurry₃` to `uncurry f = (↿ f) ∘ (finThreeArrowEquiv α)`; lean throws the error `failed to synthesize instance HasUncurry (OfArity α β 3) ?m.12793 ?m.12795`. Since my intent with `uncurry₃` is to provide specifications for cases where the types may be pretty complex (like these heterogeneous functions or the ternary functor uncurrying from the other PR) it needs to have a concrete type for unification

File: Mathlib/Logic/Function/OfHArity.lean [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+-/
+
+universe u
+
+namespace Function
+
+open Matrix (vecCons vecHead vecTail vecEmpty)
+
+/-- The type of `n`-ary functions `p 0 → p 1 → ... → p (n - 1) → τ`. -/
+def OfHArity : {n : ℕ} → (Fin n → Type u) → Type u → Type u
```
Comment (eric-wieser) [inline Mathlib/Logic/Function/OfHArity.lean:26]: I think this should perhaps be called `OfTypeTuple` or something? `OfArity` is so named because `OfArity A B 3` is a function of arity `3`. This declaration reads as `OfHArity ![B, C] D`, which doesn't actually mention the arity at all.

File: Mathlib/Data/Fin/Tuple/Curry.lean [hunk 2]
```diff
@@ -17,55 +18,147 @@ n-ary generalizations of the binary `curry` and `uncurry`.
...
-variable {α β : Type u}
+namespace Function
 
-namespace Function.OfArity
+/-- Currying for 3-ary functions. -/
+@[inline] def curry₃ {α : Type u} {β : Type v} {γ : Type w} {δ : Type w'} :
+    (α × β × γ → δ) → α → β → γ → δ := fun f a b c => f (a, b, c)
+/-- Uncurrying for 3-ary functions. -/
+@[inline] def uncurry₃ {α : Type u} {β : Type v} {γ : Type w} {δ : Type w'} :
+    (α → β → γ → δ) → (α × β × γ) → δ := fun f x => f x.1 x.2.1 x.2.2
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/Curry.lean:37]: I think these should be reviewed as a separate PR.

File: Mathlib/Logic/Function/OfArity.lean [hunk 3]
```diff
@@ -26,49 +26,56 @@ namespace Function
 
 Note that this is not universe polymorphic, as this would require that when `n=0` we produce either
 `Unit → β` or `ULift β`. -/
-def OfArity (α β : Type u) : ℕ → Type u
-  | 0 => β
-  | n + 1 => α → OfArity α β n
+def OfArity (α β : Type u) (n : ℕ) : Type u := OfHArity (fun (_ : Fin n) => α) β
```
Comment (eric-wieser) [inline Mathlib/Logic/Function/OfArity.lean:29]: I wonder if we can get away with making this an `abbrev`; does anything break?

File: Mathlib/Data/Fin/Tuple/Curry.lean (hunk 2 above)
Comment (Shamrock-Frost) [reply Mathlib/Data/Fin/Tuple/Curry.lean:37]: Sorry, I misunderstood what you were suggesting I take out from the previous PR. I'll strip the 3-currying stuff

File: Mathlib/Logic/Function/OfHArity.lean (hunk 1 above)
Comment (Shamrock-Frost) [reply Mathlib/Logic/Function/OfHArity.lean:26]: I guess I what I had in mind is that `![B, C]` is the arity of the function, eg in the sense that a multi-sorted theory has function symbols whose arities look like `![A1,...,An] -> B`. But that interpretation is pretty unclear from just the code/docs, and not that standard to begin with. The name `OfTypeTuple` also feels unclear to me, what about `FromTypes ![B, C] D`?

File: Mathlib/Logic/Function/OfArity.lean (hunk 3 above)
Comment (Shamrock-Frost) [reply Mathlib/Logic/Function/OfArity.lean:29]: Seems like it's fine, the ZFC.Basic file doesn't break in any new way

File: Mathlib/Logic/Function/OfHArity.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Logic/Function/OfHArity.lean]: `FromTypes` sounds fine too. Maybe ask on Zulip for a name suggestion?

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

File: Mathlib/Data/Fin/Tuple/Curry.lean [hunk 4]
```diff
@@ -17,55 +18,120 @@ n-ary generalizations of the binary `curry` and `uncurry`.
...
   left_inv := uncurry_curry
   right_inv := curry_uncurry
 
+lemma curry_2_eq_curry {p : Fin 2 → Type u} {τ : Type u}
+    (f : ((i : Fin 2) → p i) → τ) :
+    curry f = Function.curry (f ∘ (piFinTwoEquiv p).symm) := rfl
+
+lemma uncurry_2_eq_uncurry (p : Fin 2 → Type u) (τ : Type u)
+    (f : Function.FromTypes p τ) :
+    uncurry f = Function.uncurry f ∘ piFinTwoEquiv p := rfl
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/Curry.lean:100]: This should be `two` not `2` in the theorem name

bors r-

bors d+

**Comment by Shamrock-Frost** [comment]
bors merge
